### PR TITLE
ticket-27: Add Split by Depth/3P/"Problem Exp" logic to DataSplitter class

### PR DESCRIPTION
**Pull request by morriscb**
Tried to hook up everything I could. I wasn't able to find where DataSplitter is instantiated and used outside of notebook so I'm putting at least this code into review with the idea that we may have to ticket some more work to hook up the new Datasplitter.

### CODE REVIEW

File: src/deepcell/data_splitter.py [hunk 1]
```diff
@@ -63,12 +70,46 @@ def __init__(self, model_inputs: List[ModelInput], train_transform=None,
...
+                self._exp_labels[idx] = self._n_depth_bins + 1
+            self._exp_ids[idx] = int(exp_meta['experiment_id'])
+            depths[idx] = exp_meta['imaging_depth']
+
+        exp_arg_sort = np.argsort(self._exp_ids)
+        self._exp_ids = self._exp_ids[exp_arg_sort]
+        self._exp_labels = self._exp_labels[exp_arg_sort]
+        depths = depths[exp_arg_sort]
+
+        sorted_args = np.argsort(depths)[self._exp_labels < 0]
```
Comment (aamster) [inline src/deepcell/data_splitter.py:104]: Please add doc that the negative labels are those that heven't been assigned yet

File: src/deepcell/data_splitter.py [hunk 2]
```diff
@@ -63,12 +70,46 @@ def __init__(self, model_inputs: List[ModelInput], train_transform=None,
...
         self._center_roi_centroid = center_roi_centroid
 
+        self._create_experiment_labels(experiment_metadatas)
+
+    def _create_experiment_labels(
+            self,
+            experiment_metadatas: List[ExperimentMetadata]):
+        """
+        """
+        self._exp_labels = np.full(len(experiment_metadatas), -1, dtype=int)
```
Comment (aamster) [inline src/deepcell/data_splitter.py:87]: Instead of setting `exp_labels` as an instance attribute, can you call this method from `get_train_test_split`? IMO this logic is local to getting the train/test split.

File: src/deepcell/data_splitter.py [hunk 3]
```diff
@@ -63,12 +70,46 @@ def __init__(self, model_inputs: List[ModelInput], train_transform=None,
...
+            depths[idx] = exp_meta['imaging_depth']
+
+        exp_arg_sort = np.argsort(self._exp_ids)
+        self._exp_ids = self._exp_ids[exp_arg_sort]
+        self._exp_labels = self._exp_labels[exp_arg_sort]
+        depths = depths[exp_arg_sort]
+
+        sorted_args = np.argsort(depths)[self._exp_labels < 0]
+        bin_size = len(sorted_args) / self._n_depth_bins
+        for idx in range(self._n_depth_bins):
```
Comment (aamster) [inline src/deepcell/data_splitter.py:106]: Can you instead use `np.digitize` here?

File: src/deepcell/data_splitter.py [hunk 4]
```diff
@@ -63,12 +70,46 @@ def __init__(self, model_inputs: List[ModelInput], train_transform=None,
...
+            min_idx = int(idx * bin_size)
+            max_idx = int((idx + 1) * bin_size)
+            if max_idx > len(sorted_args):
+                max_idx = len(sorted_args)
+            self._exp_labels[sorted_args[min_idx:max_idx]] = idx
+
     def get_train_test_split(self, test_size):
```
Comment (aamster) [inline src/deepcell/data_splitter.py:113]: Please add doc for how we are splitting the data.

File: src/deepcell/data_splitter.py [hunk 5]
```diff
@@ -170,7 +211,11 @@ def _sample_dataset(self, dataset: List[ModelInput],
                 self._center_roi_centroid is True or \
                 self._center_roi_centroid == 'test'
 
-        artifacts = [dataset[i] for i in index]
+        selected_experiments = self._exp_ids[index]
+        artifacts = []
```
Comment (aamster) [inline src/deepcell/data_splitter.py:215]: Can this logic to pull all `ModelInput` that belong to `seleceted_experiments` happen outside this method? Because of the decision to not split by experiment for train/val split, this bit of code will be confusing in that case.

**Review by aamster — CHANGES_REQUESTED** [review]
The splitting logic looks good. 

I'd like to only touch train/test splitting logic, as I'm not concerned about train/val splitting logic, and the only requirement from the science group is that we split train/test this way. There is also an error in the train/val splitting logic. So I'd like to keep this logic local to train/test split only. 

This splitting logic will then only happen for train/test split. In my mind, this will happen outside of the code in a script. Therefore it doesn't need/shouldn't be added to the CLI.

File: src/deepcell/data_splitter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:90]: Changed to `_exp_bin_ids`

File: src/deepcell/data_splitter.py [hunk 6]
```diff
@@ -63,12 +70,46 @@ def __init__(self, model_inputs: List[ModelInput], train_transform=None,
         self._use_correlation_projection = use_correlation_projection
         self._centroid_brightness_quantile = centroid_brightness_quantile
         self._centroid_use_mask = centroid_use_mask
+        self._n_depth_bins = n_depth_bins
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:76]: Okay, how do we want this info sent to the splitter then?

File: src/deepcell/data_splitter.py [hunk 7]
```diff
@@ -77,10 +118,11 @@ def get_train_test_split(self, test_size):
             image_dim=self.image_dim,
             use_correlation_projection=self._use_correlation_projection
         )
-        sss = StratifiedShuffleSplit(n_splits=2, test_size=test_size,
+        sss = StratifiedShuffleSplit(n_splits=1,
+                                     test_size=test_size,
                                      random_state=self.seed)
-        train_index, test_index = next(sss.split(np.zeros(len(full_dataset)),
-                                                 full_dataset.y))
+        train_index, test_index = next(sss.split(self._exp_labels,
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:127]: Yeah, took the idea that we wanted to split by experiment and not ROI to mean that we wanted the indices to be for experiment and not the ROIs. This means that I use the sample data function to do the mapping of experiment_index -> ROI.

File: src/deepcell/data_splitter.py [hunk 8]
```diff
@@ -63,12 +70,46 @@ def __init__(self, model_inputs: List[ModelInput], train_transform=None,
...
+        """
+        """
+        self._exp_labels = np.full(len(experiment_metadatas), -1, dtype=int)
+        self._exp_ids = np.empty(len(experiment_metadatas), dtype=np.uint64)
+        depths = np.empty(len(experiment_metadatas), dtype=int)
+
+        for idx, exp_meta in enumerate(experiment_metadatas):
+            if exp_meta['problem_experiment']:
+                self._exp_labels[idx] = self._n_depth_bins
+            elif exp_meta['equipment'].startswith('3P'):
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:97]: By that do you mean we input the equipment name into the class to use here? Perhaps create some other property that comes into the metadata?

File: src/deepcell/data_splitter.py [hunk 9]
```diff
@@ -116,8 +158,8 @@ def get_cross_val_split(self, train_dataset: RoiDataset, n_splits=5,
                 is_train=False)
             yield train, valid
 
-    @staticmethod
     def get_cross_val_split_idxs(
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:164]: Right, only noticed this as I made the pull request. I think this goes with keeping the val/train split on ROI so I can back this off.

File: src/deepcell/data_splitter.py [hunk 10]
```diff
@@ -136,11 +178,10 @@ def get_cross_val_split_idxs(
         -------
         yields Tuple of np.ndarray for train and test indices
         """
-        n = len(model_inputs)
-        y = RoiDataset.get_numeric_labels(model_inputs=model_inputs)
         skf = StratifiedKFold(n_splits=n_splits, shuffle=shuffle,
                               random_state=seed)
-        for train_index, test_index in skf.split(np.zeros(n), y):
+        for train_index, test_index in skf.split(self._exp_labels,
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:186]: I think this will mean we have to write two `sample_data` methods. One for indexed by experiment and one indexed by ROI (aka the previous logic).

File: src/deepcell/cli/schemas/base.py [hunk 11]
```diff
@@ -27,6 +28,11 @@ class BaseSchema(argschema.ArgSchema):
                     'continue training using this checkpoint. If inference, '
                     'will use this checkpoint for inference'
     )
+    experiment_metadata_path = argschema.fields.InputFile(
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/cli/schemas/base.py:31]: Okay, thanks. That wasn't clear. Should I just remove all of the CLI changes I made and then just assume we are going to use DataSplitter with an input json that we just create in a notebook or something?

File: src/deepcell/data_splitter.py [hunk 12]
```diff
@@ -5,25 +5,31 @@
...
+                 train_transform=None,
                  test_transform=None, seed=None,
                  cre_line=None, exclude_mask=False,
                  mask_out_projections=False, image_dim=(128, 128),
                  use_correlation_projection=True,
                  center_roi_centroid=False,
                  centroid_brightness_quantile=0.8,
-                 centroid_use_mask=False):
+                 centroid_use_mask=False,
+                 n_depth_bins=4):
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:24]: Makes sense. So most of the methods we use then could be class methods instead of using self?

File: src/deepcell/data_splitter.py [hunk 13]
```diff
@@ -63,12 +70,46 @@ def __init__(self, model_inputs: List[ModelInput], train_transform=None,
...
 
+        self._create_experiment_labels(experiment_metadatas)
+
+    def _create_experiment_labels(
+            self,
+            experiment_metadatas: List[ExperimentMetadata]):
+        """
+        """
+        self._exp_labels = np.full(len(experiment_metadatas), -1, dtype=int)
+        self._exp_ids = np.empty(len(experiment_metadatas), dtype=np.uint64)
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:91]: Right, I we could get this from the `ModelInput` set and use, for instance `np.unique` to trim it down. I've removed these `self` variables so I don't think this is much of a issue.

File: src/deepcell/data_splitter.py (hunk 3 above)
Comment (morriscb) [reply src/deepcell/data_splitter.py:109]: I thought about using digitize but then didn't for some reason. I'll look into it again.

File: src/deepcell/data_splitter.py (hunk 5 above)
Comment (morriscb) [reply src/deepcell/data_splitter.py:218]: So convert the experiment train/test indexes into ROI train/test indexes and input them into an unmodified `_sample_dataset` method?

File: src/deepcell/data_splitter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aamster) [reply src/deepcell/data_splitter.py:76]: as arguments to `get_train_test_split`

File: src/deepcell/data_splitter.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (aamster) [reply src/deepcell/data_splitter.py:97]: The reason that the 3p data is separated is because there is only 1 experiment at each depth rather than 5. Is there a way to encode this logic into the algorithm without hardcoding `3P`?

File: src/deepcell/data_splitter.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (aamster) [reply src/deepcell/data_splitter.py:186]: I'm not sure I follow that. If the logic stays as is, why would you need 2 `sample_data` methods?

File: src/deepcell/cli/schemas/base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (aamster) [reply src/deepcell/cli/schemas/base.py:31]: Thanks, that sounds fine since we will be instantiating the `DataSplitter` via a script (to be written)

File: src/deepcell/data_splitter.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (aamster) [reply src/deepcell/data_splitter.py:24]: It seems that way, yes

File: src/deepcell/data_splitter.py (hunk 5 above)
Comment (aamster) [reply src/deepcell/data_splitter.py:218]: I think so. Just construct `dataset` within `get_train_test_split` prior to passing it to this method, so that `_sample_dataset` can stay unmodified.

File: src/deepcell/data_splitter.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py]: Because in one case the index is on the experiment and in the other the index is on the ROI. I'll have to at least put in a bit of code to convert the indices so we can reuse the _sample method for the train/test split.

File: src/deepcell/cli/schemas/base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/cli/schemas/base.py:31]: Just to confirm, I should keep the ExperimentMetadataSchema class so we can confirm the data going in?

File: src/deepcell/data_splitter.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (aamster) [reply src/deepcell/data_splitter.py]: There should only be 1 sample method I think. Just pass in indices at the ROI level rather than at the experiment level.

File: src/deepcell/cli/schemas/base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (aamster) [reply src/deepcell/cli/schemas/base.py:31]: I like that design, yes, since I don't like passing around arbitrary dictionaries.

**Comment by morriscb** [comment]
Mostly ready for another review. I'm planning to put together a unittest for the experiment binning logic. Regarding the depth binning and 3P: I think we are stuck with the 3P specific split for now based on how the test data is chunked up. If we end up labeling more, perhabs we can drop it and just let depth sampling handle it. As for the depth sampling, as discussed I kind of wanted to do a sampling that was equal number per bin in depth guaranteed. It might be a little more safe to do if there ever end up being weird distributions in depth. That said, it might be okay and safe to do a `np.linspace` binning from min->max in depth and then use digitize. I can check that `min!=max` but can't guarantee that the binning would be safe for the StratifiedShuffle split given an arbitrary depth binning. Let me know which you prefer.

**Review by aamster — APPROVED** [review]
Looks good, thanks. 
For completeness, I looked into using `np.digitize` instead, and removing the binning on hardcoding of `3P`. 

here is my code:
```
def get_depth_cutoffs(depths, n_depth_bins):
    cutoffs = []
    unique_depths = np.unique(depths)
    step_size = int(len(unique_depths) / n_depth_bins)
    for i in range(step_size, len(unique_depths)-1, step_size):
        cutoffs.append(unique_depths[i])
    return cutoffs

    depth_cutoffs = get_depth_cutoffs(depths=depths, n_depth_bins=4)
    depth_bins = np.digitize(depths, depth_cutoffs)

    depth_bin_map = dict(zip(depths, depth_bins))

    sss = StratifiedShuffleSplit(n_splits=1,
                                 test_size=0.3,
                                 random_state=1234)
    train_index, test_index = next(sss.split(depth_bins,
                                             depth_bins))

    train = depths[train_index]
    test = depths[test_index]

    df = pd.concat(
        [pd.DataFrame({'depth': train, 'dataset': 'train'}),
         pd.DataFrame({'depth': test, 'dataset': 'test'})], ignore_index=True)
    df['bin'] = df['depth'].map(depth_bin_map)
    plot_dist(df, title='Using np.digitize')

def get_split_1(depths):
    depth_cutoffs = get_depth_cutoffs(depths=depths, n_depth_bins=4)
    depth_bins = np.digitize(depths, depth_cutoffs)

    depth_bin_map = dict(zip(depths, depth_bins))

    sss = StratifiedShuffleSplit(n_splits=1,
                                 test_size=0.3,
                                 random_state=1234)
    train_index, test_index = next(sss.split(depth_bins,
                                             depth_bins))

    train = depths[train_index]
    test = depths[test_index]

    df = pd.concat(
        [pd.DataFrame({'depth': train, 'dataset': 'train'}),
         pd.DataFrame({'depth': test, 'dataset': 'test'})], ignore_index=True)
    df['bin'] = df['depth'].map(depth_bin_map)
    plot_dist(df, title='Using np.digitize')

def plot_dist(df, title):
    fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(10, 5))
    sns.histplot(
        x='depth',
        data=df,
        ax=axes[0],
        stat='probability',
        hue='bin',
        # multiple='dodge',
    )
    sns.histplot(
        x='depth',
        data=df[df['dataset'] == 'train'],
        ax=axes[1],
        stat='probability',
        hue='bin',
        # multiple='dodge',
    )
    sns.histplot(
        x='depth',
        data=df[df['dataset'] == 'test'],
        ax=axes[2],
        stat='probability',
        hue='bin',
        # multiple='dodge',
    )
    axes[0].set_title('all')
    axes[1].set_title('train')
    axes[2].set_title('test')

    plt.suptitle(title)
    plt.tight_layout()
    plt.show()
```

Here is my comparison of this approach vs the existing (without the "problem experiment" designation, since I couldn't find this list). I don't see any major differences between the distributions. In the "digitize" approach it is not hardcoding a bin for "3P" and is just using its depth (since the depth is greater than all the others)

digitize

equal_sized

File: src/deepcell/data_splitter.py [hunk 14]
```diff
@@ -65,33 +71,143 @@ def __init__(self, model_inputs: List[ModelInput], train_transform=None,
...
+        exp_bin_ids : numpy.ndarray
+           Bin assigned to each experiment id.
+        """
+        # Create our arrays we will use to bin. -1 for a bin_id denotes
+        # a experiment that has not been assigned to a bit yet.
+        exp_bin_ids = np.full(len(experiment_metadatas), -1, dtype=int)
+        depths = np.empty(len(experiment_metadatas), dtype=int)
+
+        for idx, exp_meta in enumerate(experiment_metadatas):
+            if exp_meta['problem_experiment']:
```
[the remark this answers is not in the document]
Comment (morriscb) [reply src/deepcell/data_splitter.py:108]: Understandable. Added this comment block.
`# Check for any "special" experiments that we want to partition
  # separately. This can be anything from labeling a set that have
  # problematic segmentation or have some other special attribute
  # we would like to sample fairly in the train/test split.`

**Comment by morriscb** [comment]
Thanks for the thorough review. I simplified the binning by depth logic to be linear spaced in depth. I've also upped the bins to 5 in depth to reflect the change from letting "3P" be a special flag and just letting it be pulled in by depth. This way the last bin will have the "3P" depths and 4 bins by depth for the rest of the data. I'll merge it once you get a change to take a look.

**Comment by aamster** [comment]
@morriscb looks great, thanks for making that simplification
